kdc/pkinit-ec.c

```diff
@@ -69,101 +73,85 @@
                        size_t *dh_gen_keylen)
 {
-    const EC_GROUP *group;
-    EC_KEY *ephemeral;
-    krb5_keyblock key;
-    krb5_error_code ret;
-    unsigned char *p;
-    size_t size;
-    int len;
-
-    *dh_gen_key = NULL;
-    *dh_gen_keylen = 0;
-    *ec_key_key = NULL;
-
-    memset(&key, 0, sizeof(key));
-
-    if (ec_key_pk == NULL) {
-        ret = KRB5KRB_ERR_GENERIC;
-        krb5_set_error_message(context, ret, "public_key");
-        return ret;
-    }
-
-    group = EC_KEY_get0_group(ec_key_pk);
-    if (group == NULL) {
-        ret = KRB5KRB_ERR_GENERIC;
-        krb5_set_error_message(context, ret, "failed to get the group of "
-                               "the client's public key");
-        return ret;
-    }
-
-    ephemeral = EC_KEY_new();
-    if (ephemeral == NULL)
-        return krb5_enomem(context);
-
-    EC_KEY_set_group(ephemeral, group);
-
-    if (EC_KEY_generate_key(ephemeral) != 1) {
-	EC_KEY_free(ephemeral);
-        return krb5_enomem(context);
-    }
-
-    size = (EC_GROUP_get_degree(group) + 7) / 8;
-    p = malloc(size);
-    if (p == NULL) {
-        EC_KEY_free(ephemeral);
-        return krb5_enomem(context);
-    }
-
-    len = ECDH_compute_key(p, size,
-                           EC_KEY_get0_public_key(ec_key_pk),
-                           ephemeral, NULL);
-    if (len <= 0) {
+    EVP_PKEY_CTX *pctx = NULL;
+    EVP_PKEY *ephemeral = NULL;
+    krb5_error_code ret = 0;
+    unsigned char *p = NULL;
+    size_t size = 0;
+
+    if (ec_key_pub == NULL)
+        /* XXX This seems like an internal error that should be impossible */
+        krb5_set_error_message(context, ret = KRB5KRB_ERR_GENERIC,
+                               "Missing client ECDH key agreement public key");
+    if (ret == 0 &&
+        (ephemeral =
+             EVP_EC_gen(OSSL_EC_curve_nid2name(NID_X9_62_prime256v1))) == NULL)
+        krb5_set_error_message(context, ret = KRB5KRB_ERR_GENERIC,
+                               "Could not generate an ECDH key agreement private key");
+    if (ret == 0 &&
+        (pctx = EVP_PKEY_CTX_new(ephemeral, NULL)) == NULL)
+        ret = krb5_enomem(context);
+    if (ret == 0 && EVP_PKEY_derive_init(pctx) != 1)
+        ret = krb5_enomem(context);
+    if (ret == 0 &&
+        EVP_PKEY_CTX_set_ecdh_kdf_type(pctx, EVP_PKEY_ECDH_KDF_NONE) != 1)
+        krb5_set_error_message(context, ret = KRB5KRB_ERR_GENERIC,
+                               "Could not generate an ECDH key agreement private key "
+                               "(EVP_PKEY_CTX_set_dh_kdf_type)");
+    if (ret == 0 &&
+        EVP_PKEY_derive_set_peer_ex(pctx, ec_key_pub, 1) == 0)
+        krb5_set_error_message(context, ret = KRB5KRB_ERR_GENERIC,
+                               "Could not generate an ECDH key agreement private key "
+                               "(EVP_PKEY_derive_set_peer_ex)");
+    if (ret == 0 &&
+        (EVP_PKEY_derive(pctx, NULL, &size) == 0 || size == 0))
```

Comment:
EVP_PKEY_derive appears to return 0 for success.  Errors are negative values.

Reply:
See above.